# ebay Kleinanzeigen Car Sales

The goal of the project is to clean and analyze a dataset that describes the purchases of cars on eBay Kleinanzeigen, a classifieds section of the German eBay website. To analyze the data, we will import pandas to first clean the dataset, and then draw conclusions from our cleaned set.
To view the dataset we will be working with, click [here](https://data.world/data-society/used-cars-data).

In [1]:
import pandas as pd 
import numpy as np

# uses read_csv function from pandas to open csv as formatted dataset
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

## Checking the Dataset

Here, by simply inputting the stored dataset `autos` and running the notebook, we are able to analyze our dataset to ensure that it was stored properly.

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

## Obtaining Basic Information

Here we use the methods `autos.info()` and `autos.head()` to print the first 5 rows of the `autos` dataset and obtain basic information about the types of data contained in each column.

In [3]:
print(autos.head())
autos.info()

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

## Column Renaming

Because the column headers were in `camelCase`instead of the conventional python `snake_case`, we will rename the names of the columns to conform to python conventions. To do this, using the method `autos.columns`, we will print the name of each column's header. After doing this, we will copy the existing column names, and adjust them as necessary to conform to the `snake_case` form.

In [4]:
autos.columns # Prints column headers before editing

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

print(autos.head()) # Prints first 5 rows to ensure column headers were changed correctly

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158 

## Further Dataset Review

Next, we will continue to take a deeper dive into the datset to see what needs to be cleaned. To do this, we will use the `autos.describe()` method to review descriptive statistics that will be used to determine which columns we'll be able to drop, which columns will require further investigation, and which data needs further cleaning.

In [6]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

The columns `'seller'` and `'offer_type'`, of the 50,000 data points, both contain 49,999 identical values, so these columns are strong candidates to be dropped, which we will do below using the `.drop()` method. The columns `'price'` and `'odometer'` can both be converted from string to integer values. First, we'll remove `'seller'` and `'offer_type'` from the dataframe.

In [7]:
autos = autos.drop(['seller','offer_type'], axis = 'columns')  # removes columns from dataframe

In [8]:
autos.columns # Prints column names to verify columns were dropped successfully

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen'],
      dtype='object')

Now that we've removed the columns, we'll now convert `'odometer'` and `'price'` to string values.

In [9]:
pd.value_counts(autos['odometer']) # Prints each value that occurs and how many times it occurs

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [10]:
pd.value_counts(autos['price']) # Prints each value that occurs and how many times it occurs

$0         1421
$500        781
$1,500      734
$2,500      643
$1,200      639
           ... 
$43,461       1
$1,698        1
$17,830       1
$8,875        1
$16,680       1
Name: price, Length: 2357, dtype: int64

In [11]:
price = autos['price'] # Assigns price column to the variable price
price = price.str.replace('$','') # Removes the dollar sign from the data string
price = price.str.replace(',','') # Removes the comma from the data string

autos['price'] = price.astype(int) # Assigns price back to column as a series of integer values
print(autos['price'].head()) # Prints the list first rows of the series so we can verify the changes were made correctly

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64


In [12]:
odometer = autos['odometer'] # Assigns odometer column to the variable odometer
odometer = odometer.str.replace('km','') # Removes km from the data string
odometer = odometer.str.replace(',','') # Removes the comma from the data string

autos['odometer'] = odometer.astype(int) # Assigns odometer back to column as a series of integer values
print(autos['odometer'].head()) # Prints the list first rows of the series so we can verify the changes were made correctly

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64


Now that we've updated the columns, we no longer have the unit of measure for the series `odometer`. To remedy this, we will include the the unit in the column name by renaming it `odometer_km` using the method `autos.rename()`.

In [13]:
autos.rename({'odometer' : 'odometer_km'}, inplace = True, axis = 'columns') # Renames odometer column to odometer_km
autos.columns # Method prints columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen'],
      dtype='object')

## Analysis of `odometer_km` and `price`

Next, we'll dive deeper into the series `odometer_km` and `price`. Our goal is to check the columns of data for any outliers. Ultimimately, we want to identify any extreme values and remove them from our data set.
First, we'll analyze the `price` series.

In [14]:
autos['price'].unique().shape # Outputs the number of unique values in the series

(2357,)

In [15]:
autos['price'].describe() # Prints a list of descriptive statistics about the series

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
top_sixty_prices = autos['price'].value_counts().sort_index(ascending = False).head(n=60) # Stores list of the 60 highest prices
bottom_twenty_prices = autos['price'].value_counts().sort_index(ascending = False).tail(n=20) # Stores list of the 20 lowest prices

print(top_sixty_prices, bottom_twenty_prices) # Prints highest and lowest prices so we can assess for outliers

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
145000      1
139997      1
137999      1
135000      1
130000      1
129000      1
128000      1
120000      2
119900      1
119500      1
116000      1
115991      1
115000      1
114400      1
109999      1
105000      2
104900      1
99900       2
99000       2
98500       1
94999       1
93911       1
93000       2
89900       1
89000       1
88900       1
86500       1
85000       1
84997       1
84000       1
Name: price, dtype: int64 35       1
30       7
29       1
25       5
20       4
18       1
17       3
15       2
14       1
13       2
12       3
11       2
10

There are 1421 free cars and 156 one dollar cars. Because those two price points account for approximately 3 percent of our total price points, and the bottom 25 percent of our standard distribution is 1,100 dollars, these two price points don't account for many of the total data points in the set, but are still capable of skewing the entire set. Therefore, we will be considering these two price points outlier totals. 
On the other hand, we have some very expensive cars. There are 11 cars that cost over 1,000,000 dollars that are in the dataset and 53 over 100,000 dollars. The top 75 percent value in the standard deviation is 7,200 dollars and the mean value is 9,840 dollars. This suggests that these high values, which are only representative of 64 out of the 50,000 cars in the dataset, are skewing the data so much that the average price point is over 7,000 dollars above our median price point. Therefore, we will designate everything over 100,000 dollars as an outlier value.

Next, we'll remove the outlier prices and assign the updated dataset to a new variable `cleaned_price`.

In [17]:
cleaned_price = autos[autos['price'].between(2,100000)] # Removes all rows with price points outside the range of $2 to $100,000
print(cleaned_price['price'].value_counts().sort_index(ascending = False)) # Double checks that we cleaned the set properly

99900    2
99000    2
98500    1
94999    1
93911    1
        ..
9        1
8        1
5        2
3        1
2        3
Name: price, Length: 2308, dtype: int64


Next, using our newly cleaned `cleaned_price` dataset, we'll follow the same process with the `odometer_km` series.

In [18]:
cleaned_price['odometer_km'].unique().shape # Outputs the number of unique values in the series

(13,)

In [19]:
cleaned_price['odometer_km'].describe() # Prints a list of descriptive statistics about the series

count     48370.000000
mean     125856.109159
std       39662.917522
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
top_five_km = cleaned_price['odometer_km'].value_counts().sort_index(ascending = False).head() # Stores list of the 5 highest odometer readings
bottom_five_km = cleaned_price['odometer_km'].value_counts().sort_index(ascending = False).tail() # Stores list of the 5 lowest odometer readings

print(top_five_km, bottom_five_km) # Prints highest and lowest prices so we can assess for outliers

150000    31304
125000     5044
100000     2106
90000      1733
80000      1413
Name: odometer_km, dtype: int64 40000    814
30000    773
20000    754
10000    247
5000     806
Name: odometer_km, dtype: int64


Of our remaining datapoints, over 60 percent of the cars sold have 150,000 km on the odometer. because of this, the median, 75 percent total, and max total are all 150,000. But because this accounts for 60 percent of all data points in the entire dataset, I do not feel that it is advisable to throw out over half of the rows of data. Therefore, I will not be removing any outliers from this series. Plus, the cars in the dataset are used, and it is very common for a used car to be sold with many miles on it.

Now that we've decided how many outliers to remove from the set, let's check our updated `'price'` series.

In [21]:
cleaned_price['price'].describe()

count    48370.000000
mean      5774.133450
std       7519.488641
min          2.000000
25%       1250.000000
50%       3000.000000
75%       7490.000000
max      99900.000000
Name: price, dtype: float64

This description makes a little more sense. The mean now falls within the 25 and 75 percent values. The mean is also comfortably above the median price point, which makes sense because the most expensive cars sold will always skew the price of the mean car value. The most expensive cars cost much more than any other car someone would buy, and that vast difference gives those cars more power to sway pricing averages.

## Date Columns

Next, we'll explore the date columns to see what distributions exist in the series `date_crawled`, `ad_created`, and `last_seen`. We want to use the method `.value_counts()` to determine the percentage of the dataset each of the dates comprises. To start, we'll review `date_crawled`

In [22]:
date_crawled = cleaned_price['date_crawled'].str[:10] # Assigns date value (first 10 indices in the string) to variable
# Below code will produce percent value of each date's occurance and sort from oldest to latest date
date_crawled.value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025346
2016-03-06    0.014079
2016-03-07    0.036076
2016-03-08    0.033264
2016-03-09    0.033058
2016-03-10    0.032231
2016-03-11    0.032624
2016-03-12    0.036965
2016-03-13    0.015671
2016-03-14    0.036655
2016-03-15    0.034257
2016-03-16    0.029522
2016-03-17    0.031631
2016-03-18    0.012880
2016-03-19    0.034774
2016-03-20    0.037813
2016-03-21    0.037296
2016-03-22    0.032851
2016-03-23    0.032272
2016-03-24    0.029398
2016-03-25    0.031569
2016-03-26    0.032231
2016-03-27    0.031114
2016-03-28    0.034836
2016-03-29    0.034153
2016-03-30    0.033699
2016-03-31    0.031817
2016-04-01    0.033699
2016-04-02    0.035518
2016-04-03    0.038578
2016-04-04    0.036510
2016-04-05    0.013066
2016-04-06    0.003163
2016-04-07    0.001385
Name: date_crawled, dtype: float64

No dates stuck out as having been crawled especially frequently. No individual date comprises more than 3.8 percent of the data. The only notable trend is that once the calender turned to April 5th, the amount of times the website was crawled dropped dramatically.

Next, we'll review the `ad_created` series.

In [23]:
ad_created = cleaned_price['ad_created'].str[:10] # Assigns date value (first 10 indices in the string) to variable
# Below code will produce percent value of each date's occurance and sort from oldest to latest date
ad_created.value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038826
2016-04-04    0.036882
2016-04-05    0.011784
2016-04-06    0.003246
2016-04-07    0.001240
Name: ad_created, Length: 76, dtype: float64

One thing that is immediately recognizable is that there are many more dates in `ad_created` than there were in `date_crawled`. It's also much more note worthy that the dates April 3rd and April 4th, 2016, accounted for over 7 percent of all data points, when the first five dates on list combined to make up .0105 percent of all the dates ads were created on. This suggests there could be a broader trend of more ads being posted on later dates in the set. To investigate this theory further, we will check the ten oldest and ten most recent dates in `ad_created`.

In [24]:
# Code will show percent share of dates in ad_created series for first ten dates
ad_created.value_counts(normalize = True, dropna = False).sort_index().head(n = 10)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
Name: ad_created, dtype: float64

In [25]:
# Code will show percent share of dates in ad_created series for last ten dates
ad_created.value_counts(normalize = True, dropna = False).sort_index().tail(n = 10)

2016-03-29    0.034071
2016-03-30    0.033533
2016-03-31    0.031859
2016-04-01    0.033678
2016-04-02    0.035208
2016-04-03    0.038826
2016-04-04    0.036882
2016-04-05    0.011784
2016-04-06    0.003246
2016-04-07    0.001240
Name: ad_created, dtype: float64

After a deeper dive, we can clearly see that there were substantially more ads created on the later dates in the series `ad_created` than there were in the early dates.

Finally, we will review the `last_seen` series using the same methodologies we used above.

In [26]:
last_seen = cleaned_price['last_seen'].str[:10] # Assigns date value (first 10 indices in the string) to variable
# Below code will produce percent value of each date's occurance and sort from oldest to latest date
last_seen.value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001075
2016-03-06    0.004342
2016-03-07    0.005417
2016-03-08    0.007381
2016-03-09    0.009613
2016-03-10    0.010626
2016-03-11    0.012384
2016-03-12    0.023816
2016-03-13    0.008869
2016-03-14    0.012632
2016-03-15    0.015878
2016-03-16    0.016436
2016-03-17    0.028096
2016-03-18    0.007319
2016-03-19    0.015836
2016-03-20    0.020633
2016-03-21    0.020612
2016-03-22    0.021398
2016-03-23    0.018586
2016-03-24    0.019723
2016-03-25    0.019206
2016-03-26    0.016767
2016-03-27    0.015609
2016-03-28    0.020881
2016-03-29    0.022349
2016-03-30    0.024767
2016-03-31    0.023858
2016-04-01    0.022886
2016-04-02    0.024850
2016-04-03    0.025202
2016-04-04    0.024519
2016-04-05    0.124809
2016-04-06    0.221625
2016-04-07    0.132003
Name: last_seen, dtype: float64

Here we can see the total number of dates closely mirrors that of `date_crawled`, the other column of the three added by the crawler. In this set, however, the distribution for when a post was last seen by the crawler skyrockets in the final three days of data. Collectively, the last three dates combine for approximately 47.8 percent of all data points in the series `last_seen`.

In addition to the previous three columns, we will also analyze the date column `registration_year`. To do this, we will use the `.describe()` method to glean descriptive statistics from the series.

In [27]:
cleaned_price['registration_year'].describe()

count    48370.000000
mean      2004.773847
std         88.818164
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Immediately we see the minimum registration year was 1000, and cars were notably not being used in the dark ages. And the max date was 9999, which also can't be right because that year hasn't happened yet. The max date, in theory, should be no greater than 2016. To investigate this further, we will see how many rows in `registration_year` have these incorrect years.

In [28]:
# Below code will produce percent value of each year's occurance for the ten oldest years sort from oldest to latest date
cleaned_price['registration_year'].value_counts(normalize = True, dropna = False).sort_index().head(n = 10)

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000083
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000083
Name: registration_year, dtype: float64

In [29]:
# Below code will produce percent value of each year's occurance for the 15 most recent years sort from oldest to latest date
cleaned_price['registration_year'].value_counts(normalize = True, dropna = False).sort_index().tail(n = 15)

2015    0.008001
2016    0.024871
2017    0.028654
2018    0.009717
2019    0.000041
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000083
5911    0.000021
6200    0.000021
8888    0.000021
9000    0.000021
9999    0.000062
Name: registration_year, dtype: float64

We found that the years pre-dating the 1900's accounted for only .0104 percent of all data points in `registration_year`, so those we can remove without much concern of impacting our dataset. The years after 2017 wound up accounting for just 1 percent of all the postings in the series, so I feel comfortable removing these points. I am opting to include the 2017 cars for two reasons. First, that single year accounts for 2.87 percent of the data, so that mean I would have to be very confident that removing that year from the listing is the right decision. And two, I'm not sure that is wrong because the model year of a car is one year older than the year itself. Basically, that means 2017 cars were released in 2016. Therefore, I feel comfortable making the assumption that at least some of the cars registered in 2017 have the correct registration year, therefore giving me enough reason to include that year in the dataset.

Now we'll remove the values from the dataset as previously outlined in the paragraph above.

In [30]:
# Removes all rows with registration years outside the upper and lower bounds
cleaned_year = cleaned_price[cleaned_price['registration_year'].between(1900,2017)]

Now we will print the frequency of each years occurance within the set to ensure that removed the correct outliers.

In [31]:
# Below code will produce percent value of each year's frequency distribution within 'registration_year'
cleaned_year['registration_year'].value_counts(normalize = True, dropna = False).sort_index()

1910    0.000084
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
          ...   
2013    0.016667
2014    0.013785
2015    0.008083
2016    0.025126
2017    0.028949
Name: registration_year, Length: 79, dtype: float64

Here we can see that we have successfully limited the bounds of `registration_year` to 1910-2017.

## Brand Analysis

To further explore the data, we will observe the price variability between car brands. To do this, we will once again use the `.value_counts()` method. We will use this method to examine the values in the series `brand` with the highest share of sales postings on the website. After deciding which brands we want to analyze, we will aggregate data for each brand into a dictionary for further analysis.

In [32]:
# Below code prints the 10 brands with the highest percent share of ad postings
cleaned_year['brand'].value_counts(normalize = True, dropna = False).sort_values(ascending = False).head(n = 10)

volkswagen       0.212874
bmw              0.108756
opel             0.108254
mercedes_benz    0.095890
audi             0.086156
ford             0.069949
renault          0.047705
peugeot          0.029638
fiat             0.025899
seat             0.018631
Name: brand, dtype: float64

The above are the top ten most common brands in our dataset. We can see that the tenth most common brand, `'seat'`, only accounts for 1.86 percent of all rows in the column `brand`. Therefore, I feel comfortable using that as the cutoff point for our analysis. We will be analyzing the top ten most popular brands in the dataset.

In [33]:
top_brands_prices = {} # initializes empty dictionary to store brand data
brands = ['volkswagen','bmw','opel','mercedes_benz','audi','ford','renault','peugeot','fiat','seat'] # Array of top ten brands to loop over

for b in brands:
    selected_rows = cleaned_year[cleaned_year['brand'] == b] # selects rows for brand currently being iterated over
    price = selected_rows['price'].mean() # stores average price of iterated brand
    top_brands_prices[b] = price # Assigns average price to dictionary with brand as key value
    
print(top_brands_prices)

{'volkswagen': 5366.103708791209, 'bmw': 8187.990397541771, 'opel': 2967.7866100713873, 'mercedes_benz': 8461.185144848618, 'audi': 9228.253333333334, 'ford': 3702.3705583756346, 'renault': 2451.3200525394045, 'peugeot': 3078.063424947146, 'fiat': 2795.5862903225807, 'seat': 4369.374439461883}


Looking at our dictionary `top_brands_prices`, there are a three brands that stand out as the most expensive. Those would be `'bmw'`, `'audi'`, and `'mercedes_benz`, with `'audi'` being the most expensive on average at 9228 dollars. The cheapest brand, `'fiat'`, is listed at an average price of 2,796 dollars.

To gain a greater understanding of the data associated with specific car brands, we will now follow the same process to create a dictionary for mileage by brand, called `top_brand_mileage`.

In [34]:
top_brand_mileage = {} # initializes empty dictionary to store brand data

for b in brands:
    selected_rows = cleaned_year[cleaned_year['brand'] == b] # selects rows for brand currently being iterated over
    mileage = selected_rows['odometer_km'].mean() # stores average mileage of iterated brand
    top_brand_mileage[b] = mileage # Assigns average mileage to dictionary with brand as key value
    
print(top_brand_mileage)

{'volkswagen': 128930.0431711146, 'bmw': 132708.85346648743, 'opel': 129424.07871888868, 'mercedes_benz': 131074.92920932258, 'audi': 129512.72727272728, 'ford': 124223.6488504031, 'renault': 128165.49912434326, 'peugeot': 127269.20366455251, 'fiat': 117314.51612903226, 'seat': 121530.26905829596}


From the above analysis, the only immediate takeaway is that cars of every brand are sold with a lot of mileage. The brand with the lowest average mileage is `'fiat'` with an average of 117314 kilometers. The brand with the highest average mileage is `'bmw'` with an average of 132708 kilometers.

## Brand Dataframe

To read all of this data easier, we will now put our dictionaries `top_brand_mileage` and `top_brands_prices` into the dataframe `top_brands`.

In [35]:
price_series = pd.Series(top_brands_prices) # Creates series from price dictionary
mileage_series = pd.Series(top_brand_mileage) # Creates series from mileage dictionary
print(price_series,'\n',mileage_series) # Prints both series with white space between them to make sure they are stored correctly

volkswagen       5366.103709
bmw              8187.990398
opel             2967.786610
mercedes_benz    8461.185145
audi             9228.253333
ford             3702.370558
renault          2451.320053
peugeot          3078.063425
fiat             2795.586290
seat             4369.374439
dtype: float64 
 volkswagen       128930.043171
bmw              132708.853466
opel             129424.078719
mercedes_benz    131074.929209
audi             129512.727273
ford             124223.648850
renault          128165.499124
peugeot          127269.203665
fiat             117314.516129
seat             121530.269058
dtype: float64


In [36]:
top_brands = pd.DataFrame(price_series) # initializes dataframe from price series
top_brands.rename({0:'average_price'}, inplace = True, axis = 'columns') # Renames column to 'average_price'
print(top_brands) # Verify dataset was created correctly

               average_price
volkswagen       5366.103709
bmw              8187.990398
opel             2967.786610
mercedes_benz    8461.185145
audi             9228.253333
ford             3702.370558
renault          2451.320053
peugeot          3078.063425
fiat             2795.586290
seat             4369.374439


In [37]:
top_brands['average_mileage_km'] = mileage_series # Adds 'average_mileage_km' column to dataset
print(top_brands) # Prints completed dataset

               average_price  average_mileage_km
volkswagen       5366.103709       128930.043171
bmw              8187.990398       132708.853466
opel             2967.786610       129424.078719
mercedes_benz    8461.185145       131074.929209
audi             9228.253333       129512.727273
ford             3702.370558       124223.648850
renault          2451.320053       128165.499124
peugeot          3078.063425       127269.203665
fiat             2795.586290       117314.516129
seat             4369.374439       121530.269058


Based on our dataset, there is no clear relation between either average mileage and car brand. And by extension, we have no evidence of the any impacts that average mileage has on price. But, since the cars all have similar wear and tear, our average price data is very controlled. This means that we can conclude that the resale value of `'audi'` is the highest of the brands we reviewed, and `'fiat'` has the lowest resale value. 

## German Translations

Now we will translate all german words to English. To do this we will select individual series, find all unique values using the method `.unique()`, and rename accordingly.

In [38]:
cleaned_year['fuel_type'].unique() # prints array of unique values in column

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [39]:
# Below code locates german words within dataset and  translates selected words
cleaned_year.loc[cleaned_year['fuel_type'] == 'benzin','fuel_type'] = 'petrol'
cleaned_year.loc[cleaned_year['fuel_type'] == 'elektro','fuel_type'] = 'electro'
cleaned_year.loc[cleaned_year['fuel_type'] == 'andere','fuel_type'] = 'other'

cleaned_year['fuel_type'].unique() # prints updated unique values to make sure we changed values correctly

/dataquest/system/env/python3/lib/python3.8/site-packages/pandas/core/indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


array(['lpg', 'petrol', 'diesel', nan, 'cng', 'hybrid', 'electro',
       'other'], dtype=object)

In [40]:
cleaned_year['unrepaired_damage'].unique() # prints array of unique values in column

array(['nein', nan, 'ja'], dtype=object)

In [41]:
# Below code locates german words within dataset and  translates selected words
cleaned_year.loc[cleaned_year['unrepaired_damage'] == 'nein','unrepaired_damage'] = 'no'
cleaned_year.loc[cleaned_year['unrepaired_damage'] == 'ja','unrepaired_damage'] = 'yes'

cleaned_year['unrepaired_damage'].unique() # prints updated unique values to make sure we changed values correctly

array(['no', nan, 'yes'], dtype=object)

In [42]:
cleaned_year['gearbox'].unique() # prints array of unique values in column

array(['manuell', 'automatik', nan], dtype=object)

In [43]:
# Below code locates german words within dataset and  translates selected words
cleaned_year.loc[cleaned_year['gearbox'] == 'manuell','gearbox'] = 'manual'
cleaned_year.loc[cleaned_year['gearbox'] == 'automatik','gearbox'] = 'automatic'

cleaned_year['gearbox'].unique() # prints updated unique values to make sure we changed values correctly

array(['manual', 'automatic', nan], dtype=object)

In [44]:
cleaned_year['vehicle_type'].unique() # prints array of unique values in column

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [45]:
# Below code locates german words within dataset and  translates selected words
cleaned_year.loc[cleaned_year['vehicle_type'] == 'kleinwagen','vehicle_type'] = 'small_car'
cleaned_year.loc[cleaned_year['vehicle_type'] == 'kombi','vehicle_type'] = 'station_wagon'
cleaned_year.loc[cleaned_year['vehicle_type'] == 'cabrio','vehicle_type'] = 'convertible'
cleaned_year.loc[cleaned_year['vehicle_type'] == 'andere','vehicle_type'] = 'other'

cleaned_year['vehicle_type'].unique() # prints updated unique values to make sure we changed values correctly

array(['bus', 'limousine', 'small_car', 'station_wagon', nan, 'coupe',
       'suv', 'convertible', 'other'], dtype=object)

In [46]:
cleaned_year['model'].unique() # prints array of unique values in column

array(['andere', '7er', 'golf', 'fortwo', 'focus', 'voyager', 'arosa',
       'megane', nan, 'a3', 'clio', 'vectra', 'scirocco', '3er', 'a4',
       '911', 'cooper', '5er', 'polo', 'e_klasse', '2_reihe', 'c_klasse',
       'corsa', 'mondeo', 'altea', 'a1', 'twingo', 'a_klasse', 'cl',
       '3_reihe', 's_klasse', 'sandero', 'passat', 'primera', 'fiesta',
       'wrangler', 'clubman', 'a6', 'transporter', 'astra', 'v40',
       'ibiza', 'micra', '1er', 'yaris', 'colt', '6_reihe', '5_reihe',
       'corolla', 'ka', 'tigra', 'punto', 'vito', 'cordoba', 'galaxy',
       '100', 'octavia', 'm_klasse', 'lupo', 'superb', 'meriva', 'c_max',
       'laguna', 'touran', '1_reihe', 'm_reihe', 'touareg', 'seicento',
       'avensis', 'vivaro', 'x_reihe', 'ducato', 'carnival', 'boxster',
       'signum', 'sharan', 'zafira', 'rav', 'a5', 'beetle', 'c_reihe',
       'phaeton', 'i_reihe', 'sl', 'insignia', 'up', 'civic', '80',
       'mx_reihe', 'omega', 'sorento', 'z_reihe', 'berlingo', 'clk',
       '

Since there's so many models in this set, for now we'll just translate `'andare'` to `'other'`.

In [47]:
# Below code locates german words within dataset and  translates selected words
cleaned_year.loc[cleaned_year['model'] == 'andere','model'] = 'other'

cleaned_year['model'].unique() # prints updated unique values to make sure we changed values correctly

array(['other', '7er', 'golf', 'fortwo', 'focus', 'voyager', 'arosa',
       'megane', nan, 'a3', 'clio', 'vectra', 'scirocco', '3er', 'a4',
       '911', 'cooper', '5er', 'polo', 'e_klasse', '2_reihe', 'c_klasse',
       'corsa', 'mondeo', 'altea', 'a1', 'twingo', 'a_klasse', 'cl',
       '3_reihe', 's_klasse', 'sandero', 'passat', 'primera', 'fiesta',
       'wrangler', 'clubman', 'a6', 'transporter', 'astra', 'v40',
       'ibiza', 'micra', '1er', 'yaris', 'colt', '6_reihe', '5_reihe',
       'corolla', 'ka', 'tigra', 'punto', 'vito', 'cordoba', 'galaxy',
       '100', 'octavia', 'm_klasse', 'lupo', 'superb', 'meriva', 'c_max',
       'laguna', 'touran', '1_reihe', 'm_reihe', 'touareg', 'seicento',
       'avensis', 'vivaro', 'x_reihe', 'ducato', 'carnival', 'boxster',
       'signum', 'sharan', 'zafira', 'rav', 'a5', 'beetle', 'c_reihe',
       'phaeton', 'i_reihe', 'sl', 'insignia', 'up', 'civic', '80',
       'mx_reihe', 'omega', 'sorento', 'z_reihe', 'berlingo', 'clk',
       's

In [48]:
cleaned_year.head() # prints top 5 rows of each column to verify we didn't miss any translations

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest   vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control            bus               2004     manual       158   
1   8500  control      limousine               1997  automatic       286   
2   8990     test      limousine               2009     manual       102   
3   4350  control      small_car               2007  automatic        71   
4   1350     test  station_wagon               2003     manual         0   

    model  odometer_km  registration_month fuel_type       brand  \
0   other       150000                   3       lpg     peugeot   
1     7er       150000                   6    petrol         bmw   
2    golf        70000                   7    petrol  volkswagen   
3  fortwo        70000                   6    petrol       smart   
4   focus       150000                   7    petrol        ford   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0                no  2016-03-26 00:00:00               0        79588   
1                no  2016-04-04 00:00:00               0        71034   
2                no  2016-03-26 00:00:00               0        35394   
3                no  2016-03-12 00:00:00               0        33729   
4                no  2016-04-01 00:00:00               0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Date String to Integer Conversions

Now that we've translated german words in the dataset, we will convert the string date values in `'ad_created'`, `'date_crawled'`, and `'last_seen'` to uniform integer values. We will do this by dropping the time signatures off the ends of the data points and remove all dashes, so each date reads as one continuous numnber.

In [49]:
# code removes time from each column containing date data
cleaned_year['ad_created'] = cleaned_year['ad_created'].str[:10]
cleaned_year['date_crawled'] = cleaned_year['date_crawled'].str[:10]
cleaned_year['last_seen'] = cleaned_year['last_seen'].str[:10]

# prints data to ensure we altered the data correctly
print(cleaned_year['ad_created'].head(n = 3),'\n',cleaned_year['date_crawled'].head(n = 3),'\n',cleaned_year['last_seen'].head(n = 3))

0    2016-03-26
1    2016-04-04
2    2016-03-26
Name: ad_created, dtype: object 
 0    2016-03-26
1    2016-04-04
2    2016-03-26
Name: date_crawled, dtype: object 
 0    2016-04-06
1    2016-04-06
2    2016-04-06
Name: last_seen, dtype: object


<ipython-input-49-f0c2a9d69874>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaned_year['ad_created'] = cleaned_year['ad_created'].str[:10]
<ipython-input-49-f0c2a9d69874>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaned_year['date_crawled'] = cleaned_year['date_crawled'].str[:10]
<ipython-input-49-f0c2a9d69874>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documenta

In [50]:
# code removes dashes from dataset
cleaned_year['ad_created'] = cleaned_year['ad_created'].str.replace('-','')
cleaned_year['date_crawled'] = cleaned_year['date_crawled'].str.replace('-','')
cleaned_year['last_seen'] = cleaned_year['last_seen'].str.replace('-','')

# prints data to ensure we altered the data correctly
print(cleaned_year['ad_created'].head(n = 3),'\n',cleaned_year['date_crawled'].head(n = 3),'\n',cleaned_year['last_seen'].head(n = 3))

0    20160326
1    20160404
2    20160326
Name: ad_created, dtype: object 
 0    20160326
1    20160404
2    20160326
Name: date_crawled, dtype: object 
 0    20160406
1    20160406
2    20160406
Name: last_seen, dtype: object


<ipython-input-50-cc77d78c5365>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaned_year['ad_created'] = cleaned_year['ad_created'].str.replace('-','')
<ipython-input-50-cc77d78c5365>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaned_year['date_crawled'] = cleaned_year['date_crawled'].str.replace('-','')
<ipython-input-50-cc77d78c5365>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the ca

In [51]:
autos_cleaned = cleaned_year # Renames dataset

## Brand/Model Combinations

Now that we've updated our dataset to improve its readability, we will continue our analysis of the car brands. We will now find the most common car brands. To do this, we will use the same array `brands` we used in our earlier analysis to find what the most common model is for those brands. We will create a dictionary `top_model_dict`, and turn it into a series `top_model` to add to our dataframe `top_brands`.

In [52]:
top_model_dict = {} # initializes dictionary to store brand data

for b in brands:
    rows = autos_cleaned[autos_cleaned['brand'] == b] # Stores rows for brand being iterated
    # Stores most common model and percentage in dictionary as value for brand key
    top_model_dict[b] = rows['model'].value_counts(normalize = True, dropna = False).sort_index().head(n = 1)

print(top_model_dict)

{'volkswagen': amarok    0.000687
Name: model, dtype: float64, 'bmw': 1er    0.101786
Name: model, dtype: float64, 'opel': agila    0.011769
Name: model, dtype: float64, 'mercedes_benz': a_klasse    0.125027
Name: model, dtype: float64, 'audi': 100    0.014061
Name: model, dtype: float64, 'ford': b_max    0.000299
Name: model, dtype: float64, 'renault': clio    0.214536
Name: model, dtype: float64, 'peugeot': 1_reihe    0.109232
Name: model, dtype: float64, 'fiat': 500    0.096774
Name: model, dtype: float64, 'seat': alhambra    0.045964
Name: model, dtype: float64}


In [53]:
top_model = pd.Series(top_model_dict) # converts dictionary to a series
top_brands['top_model'] = top_model # adds column to dataframe
print(top_brands) # Prints updated dataframe

               average_price  average_mileage_km  \
volkswagen       5366.103709       128930.043171   
bmw              8187.990398       132708.853466   
opel             2967.786610       129424.078719   
mercedes_benz    8461.185145       131074.929209   
audi             9228.253333       129512.727273   
ford             3702.370558       124223.648850   
renault          2451.320053       128165.499124   
peugeot          3078.063425       127269.203665   
fiat             2795.586290       117314.516129   
seat             4369.374439       121530.269058   

                                                      top_model  
volkswagen       amarok    0.000687
Name: model, dtype: float64  
bmw                 1er    0.101786
Name: model, dtype: float64  
opel              agila    0.011769
Name: model, dtype: float64  
mercedes_benz  a_klasse    0.125027
Name: model, dtype: float64  
audi                100    0.014061
Name: model, dtype: float64  
ford              b_max    0.00

After updating our dataframe, we can see that some brands have one model that dominates the market and other brands that have many models generating ads. `'renault'` is the brand that has one model with the highest percent share of all cars, as the model `'clio'` makes up 21.5 percent of all `'renault'` cars. `'ford'`, on the other hand, has the top model `'b_max'` which only accounts for .03 percent of all cars in the set. Therefore we can conclude that `'clio'` is the most popular model in the entire dataset and there isn't a clear pattern from brand to brand of one model holding the majority of all ad spaces on website.

## Mileage vs. Price

We will now further analyze the `'odometer_km'` series to see the effects of mileage on the price of the car.

In [54]:
mileage_dict = {} # Intitializes dictionary to store mileages
mileages = autos_cleaned['odometer_km'].unique() # Stores all unique mileage values

for m in mileages:
    rows = autos_cleaned[autos_cleaned['odometer_km'] == m] # stores rows for mileage being analyzed
    avg_price = rows['price'].mean() # Stores average price for each mileage
    mileage_dict[m] = avg_price # adds mileage key and price value to dictionary
    
mileage_series = pd.Series(mileage_dict)
print(mileage_series) # Displays our new series

150000     3734.544304
70000     10892.997510
50000     13497.814556
80000      9580.779685
10000     19054.683333
30000     15547.701823
125000     6129.971594
90000      8397.668028
20000     16409.695187
60000     12008.520070
5000       6510.686701
100000     7885.146283
40000     15307.757090
dtype: float64


Looking through our new series, we can see that there is a general trend of higher mileages resulting in lower sales prices. With the exception of 5000 km, the 5 mileage totals associated with the 6 lowest average price are the five highest mileage totals, 150,000, 125,000, 100,000, 90,000, and 80,000. The mileage with the highest average price point per car is 1,000 km, the second lowest mileage total. Therefore, we can conclude that mileage is a key factor in determining car value.

## Damaged vs. Non-Damaged Cars

Now we will analyze our dataset to see how damages impact a car's price point.

In [55]:
damaged_price_dict = {} # initializes dictionary
damages = autos_cleaned['unrepaired_damage'].dropna().unique() # Stores all unique numeric values in series

for d in damages:
    rows = autos_cleaned[autos_cleaned['unrepaired_damage'] == d] # stores rows being iterated over
    avg_price = rows['price'].mean() # Stores average price as a variable
    damaged_price_dict[d] = avg_price # adds damage key and price value to dictionary
    
damaged_price = pd.Series(damaged_price_dict) # converts dictionary to a series
print(damaged_price)

no     6939.08382
yes    2231.70172
dtype: float64


We can see from the above analysis that damages decrease the value of a car by over a factor of three. Therefore, we can conclude that damages are a very important factor in determining a car's value.